### ISSUE TITLE
Missing check for unneeded clone

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/rust-clippy#12204]
### Summary

A possible optimization for the `HashMap`/`HashSet` entry API is not detected currently.

### Reproducer

I tried this code:

```rust
let mut expr_cache: HashMap<String, String> = HashMap::new();
let mut expression = String::new();

// This `clone()` call is not needed!
let id = match expr_cache.entry(expression.clone()) {
    Entry::Occupied(e) => {},
    Entry::Vacant(e) => {
        // Could be `e.key()` instead of `expression`.
        buf_expr.write(expression);
    }
};
```

I expected to see this happen:

Instead, this happened:

### Version

```text
rustc 1.78.0-nightly
```
